sequences/seqs.gobra

```diff
@@ -0,0 +1,458 @@
+ghost
+requires len(xs) > 0
+ensures forall k int :: { k in xs } (k in xs) ==> result <= k
+ensures result in xs
+decreases xs
+pure func Min(xs seq[int]) (result int) {
+	return let _ := DropFirstPrependFirst(xs) in
+	(IsSingleton(xs) ? First(xs) : math.Min(First(xs), Min(DropFirst(xs))))
+}
+
+// The minimum of the concatenation of two non-empty sequences is less than or
+// equal to the minima of its two non-empty subsequences.
+ghost
+requires len(xs) > 0 && len(ys) > 0
+ensures Min(xs ++ ys) <= Min(xs)
+ensures Min(xs ++ ys) <= Min(ys)
+ensures forall k int :: { k in xs ++ ys } (k in xs ++ ys) ==> Min(xs ++ ys) <= k
+decreases xs
+pure func MinOfConcat(xs, ys seq[int]) utils.Unit {
+	return let _ := ConcatDropFirst(xs, ys) in
+	(IsSingleton(xs) ? utils.Unit{} : MinOfConcat(DropFirst(xs), ys))
+}
+
+// The maximum element in a non-empty sequence is greater than or equal to the
+// maxima of its non-empty subsequences.
+ghost
+requires 0 <= a && a < b && b <= len(xs)
+ensures Max(xs[a:b]) <= Max(xs)
+decreases
+pure func MaxOfSubseq(xs seq[int], a, b int) utils.Unit {
+	return utils.Asserting(Max(xs[a:b]) in xs)
+}
+
+// The minimum element in a non-empty sequence is less than or equal to the
+// minima of its non-empty subsequences.
+ghost
+requires 0 <= a && a < b && b <= len(xs)
+ensures Min(xs[a:b]) >= Min(xs)
+decreases
+pure func MinOfSubseq(xs seq[int], a, b int) utils.Unit {
+	return utils.Asserting(Min(xs[a:b]) in xs)
+}
+
+// TODO In case we implement generic sequences somehow, update the following
+// definitions to leverage those (e.g., generic First)
+
+// Flattens a sequence of sequences into a single sequence by concatenating
+// subsequences, starting from the first element.
+ghost
+decreases xs
+pure func Flatten(xs seq[seq[int]]) seq[int] {
+	return len(xs) == 0 ? Empty() : xs[0] ++ Flatten(xs[1:])
+}
+
+// Flattening sequences of sequences if distributive over concatenation.
+ghost
+ensures Flatten(xs ++ ys) == Flatten(xs) ++ Flatten(ys)
+decreases xs
+pure func FlattenConcat(xs, ys seq[seq[int]]) utils.Unit {
+	return len(xs) == 0 ? utils.Unit{} :
+	let _ := utils.Asserting((xs ++ ys)[1:] == xs[1:] ++ ys) in
+	FlattenConcat(xs[1:], ys)
+}
+
+// Flattens a sequence of sequences into a single sequence by concatenating
+// subsequences in reverse order, i.e. starting from the last element.
+// QUES With this description I thought [[1, 2], [3, 4]] gets turned into
+// [3, 4, 1, 2]. But this is not what happens; instead, it gets turned into
+// [1, 2, 3, 4]. How should we make this clear?
+ghost
+decreases xs
+// QUES Dafny calls this FlattenReverse. I think that is a bit confusing, as it
+// could mean we first Flatten, and then Reverse, which is not the same. Which
+// name should we choose? Or maybe there is a different name that is better?
+pure func ReversedFlatten(xs seq[seq[int]]) seq[int] {
+	return len(xs) == 0 ? Empty() : ReversedFlatten(xs[:len(xs)-1]) ++ xs[len(xs)-1]
+}
```

Comment:
I would drop this for now, I don't see the need for it and, as you pointed out, causes unnecessary confusion